agents/ModularAgent.py

```diff
@@ -0,0 +1,649 @@
+		return tuple(unground(e) for e in arg)
+	elif isinstance(arg, str):
+		return arg.replace('QM', '?')
+	else:
+		return arg
+
+def replace_vars(arg, i=0):
+	"""
+	Doc String
+	"""
+	if isinstance(arg, tuple):
+		ret = []
+		for elem in arg:
+			replaced, i = replace_vars(elem, i)
+			ret.append(replaced)
+		return tuple(ret), i
+	elif isinstance(arg, str) and len(arg) > 0 and arg[0] == '?':
+		return '?foa%s' % (str(i)), i+1
+	else:
+		return arg, i
+
+def variablize_by_where(state,match):
+	# print("MATCH FOUND", skill_label, exp, match)
+	
+	# vmapping = {'?foa' + str(i): ele for i, ele in enumerate(match)}
+	# r_exp = list(rename_flat({exp: True}, vmapping))[0]
+
+	mapping = {'foa' + str(i): ele for i, ele in enumerate(match)}
+	r_state = rename_flat(state, {mapping[a]: a for a in mapping})
+	return r_state
+
+
+EMPTY_RESPONSE = {}
+
+class ModularAgent(BaseAgent):
+
+	def __init__(self, feature_set, function_set, 
+									 when_learner='trestle', where_learner='MostSpecific',
+									 heuristic_learner='proportion_correct', how_cull_rule='most_parsimonious',
+									 search_depth=1, numerical_epsilon=0.0):
+		self.where_learner = get_where_learner(where_learner)
+		self.when_learner = get_when_learner(when_learner)
+		self.which_learner = get_which_learner(heuristic_learner,how_cull_rule)
+		self.rhs_list = []
+		self.rhs_by_label = {}
+		self.rhs_by_how = {}
+		# self.examples = {}
+		self.feature_set = feature_set
+		self.function_set = function_set
+		self.search_depth = search_depth
+		self.epsilon = numerical_epsilon
+		self.rhs_counter = 0
+
+
+	#######-----------------------MISC----------------------------##########
+	def apply_featureset(self, state):
+		####----------Definitely Vectorizable -------------######
+		tup = Tuplizer()
+		flt = Flattener()
+		state = flt.transform(tup.transform(state))
+
+		knowledge_base = FoPlanner([(ground(a), state[a].replace('?', 'QM')
+									 if isinstance(state[a], str)
+									 else state[a])
+									for a in state],
+								   self.feature_set)
+		knowledge_base.fc_infer(depth=1, epsilon=self.epsilon)
+		state = {unground(a): v.replace("QM", "?")
+				 if isinstance(v, str)
+				 else v
+				 for a, v in knowledge_base.facts}
+		return state,knowledge_base
+
+
+	#######---------------Needs to be implemented elsewhere--------------------##########
+	# def when_learner.applicable_skills(state,skills=None):
+	# 	pass
+	# def which_learner.most_relevant(explanations):
+	# 	pass
+	# def which_learner.cull_how(explanations):
+	# 	pass
+	# def where_learner.check_match(explanation):
+	# 	pass
+	# def how_learner.how_search(state,sai):
+	# 	pass
+
+	#####-----------------------REQUEST--------------------------###########
+
+	def applicable_explanations(self,state,rhs_list=None):# -> returns Iterator<Explanation>
+		if(rhs_list == None): rhs_list = self.rhs_list
+
+		#order here might need to be switched depending on when learner implementation (see nitty gritty questions below)
```

Comment:
I am starting to think that this idea of mine of having the when get checked before where is ill-conceived simply because the selection is not yet chosen. So yes this is resolved and probably this line will ultimately need to be removed.
if(self.when_learner.state_format == "state_only"):
			rhs_list = self.when_learner.applicable_RHSs(state, rhs_list=rhs_list)

Reply:
Yes, you absolutely need to check where before when can be checked. Mainly because when depends on the elements being renamed based on the where match.

Reply:
✓